Allow disabling per-Gen hardware support; add CI testing

mattst88 opened the pull request:
Commit messages contain explanations and data.

tl;dr: This allows the `-DGEN*=OFF` options to work by ignoring them in cases where we know the build is broken. That allows us to use them to disable the large per-Gen media kernels, which are 4.36 MiB for Gen12, 3.77 MiB for Gen11, 3.86 MiB for Gen10, 4.85 MiB for Gen9, 2.47 MiB for Gen8.

CI builds are added so we don't regress. Two follow-on patches disable Gen10 by default (since it never shipped), and (re)disable the Gen8 driver code now that PR 1302 has landed.

cc: @edwarddavidbaker in case you're interested.

mattst88 commented:
*tap* *tap*

Is this thing on?

XinfengZhang commented on .github/workflows/ubuntu.yml:
```diff
@@ -336,3 +336,328 @@ jobs:
+        make VERBOSE=1 -j$(nproc)
+        sudo make install
+
+  gcc-10-gen10-on:
```
suggest to remove this one, we will disable gen10 by default.

mattst88 replied:
I reversed the condition from -DGEN10=OFF to ON in this CI config in the patch that disables Gen10 by default.

What is it that you would rather I do?

XinfengZhang replied:
I mean we may do not need this catalog "gcc-10-gen10-on" or "gcc-10-gen10-off" in near future. of course ,we may keep it during transition period

mattst88 replied:
Right, I totally agree.

XinfengZhang commented on .github/workflows/ubuntu.yml:
```diff
@@ -336,3 +336,328 @@ jobs:
+        make VERBOSE=1 -j$(nproc)
+        sudo make install
+
+  gcc-10-gen11-off:
```
also, suppose need correlated changes if Gen10 is disabled.

mattst88 replied:
I don't understand what you mean. What change to the Gen11 CI config should I make when disabling Gen10 by default?

Enable Gen10 in the Gen11 CI config? Why?

mattst88 replied:
Okay. I'm resolving this conversation -- unless I've misunderstood.

XinfengZhang commented on media_driver/cmake/linux/media_gen_flags_linux.cmake:
```diff
@@ -18,6 +18,8 @@
 # ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 # OTHER DEALINGS IN THE SOFTWARE.
 
+option(BROKEN "Accept potentially broken builds when disabling generation-specific options" OFF)
```
BROKEN=ON, then could enable per-gen build. what's the motivation about this build option, why GENXX = OFF is not enough?

mattst88 replied:
Currently the per-Gen build options are broken, except for Gen8.

But the per-Gen build options control two separate things: 1. The per-Gen media kernel binaries, and 2. The per-Gen driver code.

The per-Gen driver code cannot be disabled arbitrarily now.

The per-Gen media kernel binaries can be disabled arbitrarily, and they are large and disabling them provides large file size reductions.

The BROKEN option controls whether the per-Gen driver code is affected by the per-Gen config options, allowing us to disable the per-Gen media kernel binaries.

dvrogozh replied:
I understand the desire to have options to reduce driver footprint, but it's not good idea to workaround something which should actually be fixed. Issue with gen10 switch off is a clear issue and needs proper resolution.

I think we should not introduce BROKEN option. Instead, I suggest to start introducing new build configurations (w/ disabled gens) which currently works to make sure we don't regress. For example, -DGEN8=OFF got fixed, let's have this path covered in github actions ci. And add other configs which currently build ok.

mattst88 replied:
I have some work in progress patches to fix Gen9, but I don't expect to be able to fix disabling the driver code for Gen10, Gen11, or Gen12.

Is someone from Intel going to sign up for that?

We can of course ship my patch as is in ChromeOS to be able to reduce the binary size, but I feel like I've provided a solid and incremental path forward that gets 90% of the size reduction, strictly improves the situation, and doesn't force anyone to drop what they're doing to untangle the currently broken per-Gen driver code disablement.

Alternatively, I'm happy to either
 -  rename the `BROKEN` option if the name is the problem
 - or I can remove `BROKEN` entirely and make the per-Gen driver code that is currently required non-optional, leaving the `GEN*` options to only disable the media kernel binaries. I create the `BROKEN` option to make the incremental work of fixing the remaining Gen options simpler.

mattst88 replied:
> * rename the BROKEN option if the name is the problem

Having received no feedback for a month, I've speculatively made this change, changing the option name to `-DONLY_MEDIA_KERNELS` and inverting the value (so that you have to set it to `OFF` to produce a broken build with `-DGEN*=OFF`).

mattst88 commented:
Anyone want to continue the discussions?

XinfengZhang commented on media_driver/cmake/linux/media_gen_flags_linux.cmake:
```diff
@@ -18,6 +18,9 @@
 # ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 # OTHER DEALINGS IN THE SOFTWARE.
 
+option(ONLY_MEDIA_KERNELS "Make per-Gen options disable only media-kernels (not driver code) if driver code is known to be required for a successful build"
```
want to clarify these:
1. it is a build option during per-gen options is not clean. related driver code still could not excluded when the per-gen build option is "OFF"
2. from previous conditions "if(ONLY_MEDIA_KERNELS OR GEN12_TGLLP);xxx"; if the option is "ON", it will always build related media driver.  your expectation should be , if ONLY_MEDIA_KERNELS is ON, but GEN12_TGGLP is OFF, then the media kernel files of TGLLP should not be compiled , but the code of TGLLP will still be compiled, right?
from this perspective, the ONLY_MEDIA_KERNELS is looks confusing, it looks not a good name to be a switch.  it is something like "allow source code compiled even disable the platform",  "ALLOW_REDUNDANT_CODE" is also not good enough, I still have no idea with a better name.  or remove the new definitions, but add comments to describe that current status, GENxx=OFF maybe just disable the kernel files.  both methods looks not so graceful. 
which configuration you cared mostly?  if you just care 1 or 2 platform disabling, maybe we could focus on that , resolve it one by one?
3. your changes with github actions does not include any reference of MEDIA_KERNELS, so, why it will benefit your project?

mattst88 replied:
>     1. it is a build option during per-gen options is not clean. related driver code still could not excluded when the per-gen build option is "OFF"

It's true that it's not a clean option. My plan/hope is for the `ONLY_MEDIA_KERNELS` option to be removed once each Gen's driver code can be safely disabled.

>     2. from previous conditions "if(ONLY_MEDIA_KERNELS OR GEN12_TGLLP);xxx"; if the option is "ON", it will always build related media driver.  your expectation should be , if ONLY_MEDIA_KERNELS is ON, but GEN12_TGGLP is OFF, then the media kernel files of TGLLP should not be compiled , but the code of TGLLP will still be compiled, right?

That's exactly right.

>        from this perspective, the ONLY_MEDIA_KERNELS is looks confusing, it looks not a good name to be a switch.  it is something like "allow source code compiled even disable the platform",  "ALLOW_REDUNDANT_CODE" is also not good enough, I still have no idea with a better name.  or remove the new definitions, but add comments to describe that current status, GENxx=OFF maybe just disable the kernel files.  both methods looks not so graceful.
>        which configuration you cared mostly?  if you just care 1 or 2 platform disabling, maybe we could focus on that , resolve it one by one?

I agree, it's not graceful and it's somewhat confusing. I'd kind of suggest that this isn't unexpected given that almost none of the `-DGEN` options currently work if set to `OFF` :)

>     3. your changes with github actions does not include any reference of MEDIA_KERNELS, so, why it will benefit your project?

It would benefit anyone that wishes to reduce the disk usage of the project by disabling media kernels (and code, in the case of Gen8).

The intention behind the `ONLY_MEDIA_KERNELS` option (which is `ON` by default) is to provide a baseline of support for disabling the `GEN` options. Currently most of them, when disabled, cause the build to fail. My strategy is to

1. avoid those build failures with the `ONLY_MEDIA_KERNELS` mechanism (while still providing a simple way to reproduce the build failures for purposes of fixing them)
2. allow significant disk usage savings by disabling the media kernels even in cases where the rest of that Gen's driver support code cannot (yet) be disabled

To that end, the GitHub Actions are there to ensure that each of the `-DGEN=` options does not produce a broken build. In the case of Gen 8, this means disabling the media kernels *and* driver code. For other Gens it means only disabling the media kernels -- until those issues are resolved which will enable us to disable the Gen-specific driver code as well.

I could invert the default of `ONLY_MEDIA_KERNELS` from `ON` to `OFF` so that a build configuration that only specified `-DGEN9=OFF` produces a build failure, and I could add `-DONLY_MEDIA_KERNELS=ON` to the GitHub Action for Gen9, but I don't know why that would be better in any way.


Could I get some clarity on what the concerns are? 

* Is the goal of building the project for a single generation in order to reduce the disk usage from 36 MiB to 18 MiB not considered sufficiently valuable?

* Is the strategy of achieving 90% of the goal by allowing disabling media kernels independently of resolving long standing issues not a good one? (Issue #451 was filed in 2018; similar issues #813 and #1065 were filed later).

I'm unsure whether we're in agreement on the goals, the strategy, both, or neither.

Like if it's just confusion about how the option should be named or how it should behave then, great, we can surely find something better. If there's disagreement about the value in reducing binary size, then we probably need to take a step back and have a different discussion. In between those, if there's disagreement on the strategy of allowing disabling media kernels independently of driver code, then we probably need to figure out a new path.

XinfengZhang replied:
thanks @mattst88,  I just noticed you did not add "ONLY_MEDIA_KERNEL" for gen8, because gen8 is clean. yes, we should not change the value of "ONLY_MEDIA_KERNEL" in  configuration Github Actions.
I have no concern with the commit itself , except the naming. better to "ALLOW_CODE_BESIDES_KERNEL" ?  , "ALLOW_REDUNDANT_CODE" or .... , just like what I said, I have no good idea about it. anyway, I agree the value and the purpose. thanks for your contribution

mattst88 replied:
Okay, thank you very much!

I've changed the name to `ENABLE_REQUIRED_GEN_CODE` which I think is better than `ONLY_MEDIA_KERNEL`. I've also added a comment above the option documenting that we plan to remove it when all Gen options can disable the driver code.

XinfengZhang commented:
2 build failure.

mattst88 commented:
Okay, rebased onto master, updated for some platform additions, and fixed the two build failures.

Looks like I need someone to approve the workflow to run the CI again.

mattst88 commented:
> Okay, rebased onto master, updated for some platform additions, and fixed the two build failures.

More fixes. Should be good to go now.

mattst88 commented:
Thank you. I'll open further PRs that allow disabling per-Gen driver code as I have time. Hopefully those will not take nearly as long to merge.
